Add OAuth SSO, SAML SSO, and General Security resources

drcrees opened the pull request:
### New Resources
- `artifactory_general_security`
- `artifactory_oauth_settings`
- `artifactory_saml_settings`

### The Unconventionals
- Terraform `update` and `delete` performed via `PATCH` YAML requests
     - `<base-url>/artifactory/api/system/configuration`
     - [JFrog Artifactory YAML Configuration Documentation](https://www.jfrog.com/confluence/display/JFROG/Artifactory+YAML+Configuration#ArtifactoryYAMLConfiguration-Security(Generalsecurity,PasswordPolicy,LDAP,SAML,OAuth,HTTPSSO,Crowd))
- Terraform `read` performed via one of the following **_undocumented_** API endpoints 
     - `<base-url>/artifactory/api/securityconfig`
     - `<base-url>/artifactory/api/oauth`
     - `<base-url>/artifactory/api/saml/config`

drcrees commented:
I have read the CLA Document and I hereby sign the CLA

drcrees commented:
recheckcla

drcrees commented:
recheckcla

drcrees commented:
@chb0github is there anything that you're waiting on or would like me to do for this?

chb0github commented:
@drcrees - Somehow I didn't even see this - let me have a look

chb0github commented on pkg/artifactory/datasource_artifactory_fileinfo.go:
```diff
@@ -3,7 +3,8 @@ package artifactory
 	"fmt"
-	"github.com/atlassian/go-artifactory/v2/artifactory/v1"
+
+	v1 "github.com/atlassian/go-artifactory/v2/artifactory/v1"
```
why?

drcrees replied:
Result of `go fmt`

chb0github commented on pkg/artifactory/resource_artifactory_group.go:
```diff
@@ -3,10 +3,11 @@ package artifactory
-	"github.com/atlassian/go-artifactory/v2/artifactory/v1"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+
+	v1 "github.com/atlassian/go-artifactory/v2/artifactory/v1"
```
???

drcrees replied:
Result of `go fmt`

chb0github commented on pkg/artifactory/resource_artifactory_saml_settings.go:
```diff
@@ -0,0 +1,203 @@
+		return fmt.Errorf("failed to retrieve data from <base_url>/artifactory/api/saml/config during Read")
+	}
+
+	err = json.Unmarshal(body, &samlSettings)
```
Ok, I am befuddled: You `GET` and it's `JSON`, but below, you `PATCH` and send it `YAML`?

drcrees replied:
`GET`
 The three undocumented API endpoints return JSON,
`<base-url>/artifactory/api/securityconfig`
`<base-url>/artifactory/api/oauth`
`<base-url>/artifactory/api/saml/config`

`PATCH` 
The documented endpoint for configuration only supports a YAML media type,
`<base-url>/artifactory/api/system/configuration`
[JFrog Artifactory YAML Configuration Documentation](https://www.jfrog.com/confluence/display/JFROG/Artifactory+YAML+Configuration#ArtifactoryYAMLConfiguration-Security(Generalsecurity,PasswordPolicy,LDAP,SAML,OAuth,HTTPSSO,Crowd))

chb0github commented:
Ok, so 2 general points here before I can proceed:

1. I see a bunch of import changes. I'd need to understand why (go fmt?)
2. Where did you find the documentation for these endpoints??

chb0github commented:
So, when I hit these endpoints on our test system, here is what I get:

```bash
christianb@christianb-mac terraform-provider-artifactory % curl -snL https://partnership.jfrog.io/artifactory/api/system/configuration
{
  "errors" : [ {
    "status" : 400,
    "message" : "Artifactory Online does not support accessing this API endpoint, please contact support@jfrog.com for further assistance if required."
  } ]
}%         
```
```bash                                                                                                                                                                                                                             christianb@christianb-mac terraform-provider-artifactory % curl -H "accept: application/json" -snL https://partnership.jfrog.io/artifactory/api/system/configuration
{
  "errors" : [ {
    "status" : 406,
    "message" : "Not Acceptable"
  } ]
}%         
```
```bash                                                                                                                                                                                                                             christianb@christianb-mac terraform-provider-artifactory % curl -H "accept: application/yaml" -snL https://partnership.jfrog.io/artifactory/api/system/configuration
{
  "errors" : [ {
    "status" : 406,
    "message" : "Not Acceptable"
  } ]
}%         
```
```bash                                                                                                                                                                                                                             christianb@christianb-mac terraform-provider-artifactory % curl -H "accept: application/xml" -snL https://partnership.jfrog.io/artifactory/api/system/configuration
{
  "errors" : [ {
    "status" : 400,
    "message" : "Artifactory Online does not support accessing this API endpoint, please contact support@jfrog.com for further assistance if required."
  } ]
}%     
```                                                                                                                                                                                                                                 

I have to see the documentation that supports this otherwise, despite the fact that I find your code to be pretty clean, I will have to decline it because it seems to be an undocumented API

drcrees commented:
Thanks for reviewing, @chb0github.

> I see a bunch of import changes. I'd need to understand why (go fmt?)

You are correct, import changes are from running `go fmt`.

> Where did you find the documentation for these endpoints??
>
> I have to see the documentation that supports this otherwise, despite the fact that I find your code to be pretty clean, I will have to decline it because it seems to be an undocumented API

As mentioned in my description, the endpoints used for **READ** operations are undocumented, or at least I haven't been able to find any documentation on them.  I simply inspected what the UI was doing to gather that information.

We are currently running Artifactory Enterprise w/ Xray 7.18.3 rev 71803900, 
and are using these changes in production for what it's worth.

Alternatively, I could parse the XML provided by the following endpoint (same endpoint used for patching),
**GET** `<base-url>/artifactory/api/system/configuration`

It's certainly doable, but not ideal since the endpoint returns a _massive_ XML config dump for the entire system.
Is there anything else you can think of that could be done here instead?

Otherwise, I can make those changes, as ugly as they may be lol

chb0github commented on pkg/artifactory/resource_artifactory_oauth_settings.go:
```diff
@@ -0,0 +1,247 @@
+	PersistUsers             bool                             `yaml:"persistUsers" json:"persistUsers"`
+	AllowUserToAccessProfile bool                             `yaml:"allowUserToAccessProfile" json:"allowUserToAccessProfile"`
+	OauthProvidersSettings   map[string]OauthProviderSettings `yaml:"oauthProvidersSettings"`
+	AvailableTypes           []OauthType                      `json:"availableTypes"`
```
??? Wait... I am baffled. Some should marshal as JSON and some as YAML? Or, is this json jammed into a yaml field?

drcrees replied:
Struct fields with the `yaml` tag are used to marshal as YAML for `PATCH` requests
Struct fields with the `json` tag are used to unmarshal from the undocumented API endpoints JSON response

All fields from the undocumented API endpoints are included, so there are extras (e.g., `AvailableTypes`)

This particular field, `AvailableTypes`, could be used to validate that only a supported OAuth provider types are set in `OauthProvidersSettings`.  That validation is not currently implemented, but could certainly be added.

chb0github replied:
What are you proposing for validation? Static validation or dynamic?

chb0github reviewed (changes requested):
So, I have a bunch of outstanding questions - since these are half-undocumented and not, and you clearly know more about them than me, then I need those questions answered so that I can understand them better

chb0github commented:
Ok, here is what I discovered: The APIs you're hitting will work with self-hosted but not saas and I can see the reasoning why, though I might not totally agree

So, here is what we can do: 
1. Everywhere you're using an undocumented API, emit a warning. Anywhere you're using a UI based endpoint, emit a warning.
2. I need a complete list of the endpoints you're hitting so I can try it against a SaaS instance I have access to. Any API that fails will need to use validation code that can check if the API can be supported. If you hit this endpoint:
```bash
curl  -snL https://partnership.jfrog.io/artifactory/api/system/licenses/ | jq .               
{
  "errors": [
    {
      "status": 400,
      "message": "Artifactory Online does not require license, Please contact support@jfrog.com for further assistance if required."
    }
  ]
}
```
then a failure seems to constitute that it's SaaS. So, if this returns an error, then it your code should fail validation

I think we can accept this code, but with a lot of warning to potential users. If I/we accept it, it's accepted as official and supported. However, if undocumented APIs are in use then supporting it might end up out my hands or my teams. Hence the safe-guards and caveats

drcrees commented:
> 1. Everywhere you're using an undocumented API, emit a warning. Anywhere you're using a UI based endpoint, emit a warning.

Sounds good.

> 2. I need a complete list of the endpoints you're hitting so I can try it against a SaaS instance I have access to. Any API that fails will need to use validation code that can check if the API can be supported. If you hit this endpoint:

Here is a list of endpoints being used, these are also available in the PR description.

`<base-url>/artifactory/api/system/configuration` - [documented](https://www.jfrog.com/confluence/display/JFROG/Artifactory+YAML+Configuration#ArtifactoryYAMLConfiguration-Security(Generalsecurity,PasswordPolicy,LDAP,SAML,OAuth,HTTPSSO,Crowd))
`<base-url>/artifactory/api/securityconfig` - undocumented
`<base-url>/artifactory/api/oauth` - undocumented
`<base-url>/artifactory/api/saml/config` - undocumented

If all my other answers to your questions make sense, 
let me know when I have the green light and I can work on these changes when I have time.  Thanks!

chb0github commented:
Yeah, I am satisfied with what I see here. Please make the changes asked. When this TF provider is talking to a SaaS instance it has to fail fast with an explanation

chb0github commented:
Any updates?

drcrees commented:
@chb0github sorry for the delay.  I will get the changes you've requested done sometime this week.

drcrees commented:
Again, sorry for the delay here.  I have not forgotten.  I should have some time to work on this shortly, been busy!

drcrees commented:
Addresses #94 

Using v2 of the `terraform-plugin-sdk` is a prerequisite to use better `diag.Diagnostics` capabilities for emitting more precise errors and warnings.

drcrees commented:
@chb0github this is ready for re-review.

After these updates I created a SaaS environment to test against (`crees.jfrog.io`).
I don't know if anything has changed, but all undocumented API endpoints we are wanting to use actually work.

These are the undocumented endpoints, and they work fine for me against `jfrog.io` SaaS:
`https://crees.jfrog.io/artifactory/api/oauth`
`https://crees.jfrog.io/artifactory/api/saml/config`
`https://crees.jfrog.io/artifactory/api/securityconfig`

I still think we should output warnings when using these, but we will need to update current verbiage if it actually works for SaaS.

There were quite a few additional changes that were needed in order to add proper warnings.  I will list and describe all of these changes below, I hope you will find them beneficial.  Also, huge thanks to @sailorgallifrey for his assistance in making a lot of these changes.  

Any changes that you see that aren't called out below, please add a comment with your questions and concerns and I will address.

## Updates
 - Added warnings as suggested, see collapsed example warning output below
      - we are only using `artifactory_general_security` and `artifactory_saml_settings` resources, but a warning will surface for the `artifactory_oauth_settings` resource as well
 - If using SaaS offering of Artifactory, when trying to use new security resources the provider will fail fast after calling out to unsupported API endpoints with an explanation as to why.  
     - **NOTE: I have not gotten an error to surface since these endpoints actually appear to be working against SaaS**
- Upgraded `terraform-plugin-sdk` to v2 (version 2.7.0)
     - addresses issue #94
     - adds many bug fixes and new features such as `diag.Diagnostics` used by new security resources in this PR
     - for a comprehensive list of changes view the official [CHANGELOG](https://github.com/hashicorp/terraform-plugin-sdk/blob/main/CHANGELOG.md)
          - all observed breaking changes to the provider itself have been addressed in this PR
          - [2.0.0-rc.1](https://github.com/hashicorp/terraform-plugin-sdk/blob/main/CHANGELOG.md#200-rc1-may-05-2020)
              - notable breaking changes:
                   - the SDK no longer supports protocol 4 (Terraform 0.11 and below). Providers built on the SDK after v2 will need Terraform 0.12 to be used
                   - `terraform.ResourceProvider` interface has been removed. The `*schema.Provider` type should be used directly instead **(addressed in this PR)**
                   - Updated type signatures for some functions to include `context.Context` support **(addressed in this PR)**
              - notable features:
                   - many functions in the SDK now return a new `diag.Diagnostics` type, like the new `CreateContextFunc`, `ReadContextFunc`, `UpdateContextFunc`, `DeleteContextFunc`, and a new `ValidateDiagFunc`. When using these Diagnostics, Terraform will now indicate more precisely-scoped errors, and providers now have the ability to display warnings
                        - this exact feature is what prompted the upgrade to v2 of the plugin and is used by the new security resources added in this PR
          - [2.0.0-rc.2](https://github.com/hashicorp/terraform-plugin-sdk/blob/main/CHANGELOG.md#200-rc1-may-05-2020)
              - notable breaking changes:
                   - running provider tests now requires Terraform 0.12.26 or higher
                   - removed `helper/hashcode` package. These packages can be easily replicated in plugin code if necessary or the v1 versions can be used side-by-side **(addressed in this PR)**

<details>
<summary>Example Warning Output for Undocumented Endpoints</summary>

![warning](https://user-images.githubusercontent.com/11447165/130126729-aacf1ac2-578b-4927-8b96-d688a47fa75e.png)

</details>

<details>
<summary>Example Error Output for Undocumented Endpoints against SaaS</summary>

I have not been able to reproduce this, please advise.  All undocumented endpoints work for me against my test SaaS environment on `jfrog.io`.

Current logic will surface a descriptive error if calls to any of these undocumented endpoints _fail_.
Since they are not failing I cannot show an example.

If these endpoints are working against SaaS now, we should just surface warnings.

</details>

chb0github commented:
Oh, man! You kill me. You're addressing a lot of things in 1 PR. Moving to TF sdkv2 is a good step but also leaves me worried something could break. I am in the middle of working on 2.3.x and making a major version bump so I'd prefer to fold your changes into that branch. 

I still get an error at this endpoint:

```bash
christianb@christianb-mac ~ % curl -snL https://partnership.jfrog.io/artifactory/api/system/configuration | jq .
{
  "errors": [
    {
      "status": 400,
      "message": "Artifactory Online does not support accessing this API endpoint, please contact support@jfrog.com for further assistance if required."
    }
  ]
}
```

I too am no longer getting any errors on the endpoints that are undocumented. But, undocumented is undocumented. 
Let's change the warning to basically say `... uses artifactory endpoints that are undocumented and may not be supported on SaaS endpoints, or may change without any notice. `

As long as it's understood that it's "caveat emptor" I am ok with that. 

I am re reviewing all your code now. I may have your merge it into branch `2.3.x` unless there is a burning need to release sooner

drcrees commented:
> I still get an error at this endpoint:
> 
> ```shell
> christianb@christianb-mac ~ % curl -snL https://partnership.jfrog.io/artifactory/api/system/configuration | jq .
> {
>   "errors": [
>     {
>       "status": 400,
>       "message": "Artifactory Online does not support accessing this API endpoint, please contact support@jfrog.com for further assistance if required."
>     }
>   ]
> }
> ```

That endpoint is documented, but looks like we cannot execute a `GET` against that endpoint for Artifactory Online. 
We do not execute a `GET` within the provider, just `PATCH` requests.

I've just confirmed that `PATCH` works for `/api/system/configuration` with `Content-Type: application/yaml` as expected against my test SaaS environment.

drcrees commented:
> Oh, man! You kill me. You're addressing a lot of things in 1 PR. Moving to TF sdkv2 is a good step but also leaves me worried something could break. 

FWIW, we're using these changes in production without issue.  Our updated acceptance test suite also passes.
We certainly don't have definitions for every resource, but that's what the acceptance tests are for, right?

> I too am no longer getting any errors on the endpoints that are undocumented. But, undocumented is undocumented.
Let's change the warning to basically say ... uses artifactory endpoints that are undocumented and may not be supported on SaaS endpoints, or may change without any notice.

Sounds good, I will update verbiage.

> I may have your merge it into branch 2.3.x unless there is a burning need to release sooner

How long until `2.3.x` is released do you think?  Our main objective is to stop using and maintaining our own fork.

chb0github commented:
I am hammering on the Keyboard furiously working on it. My target is before the end of the month or very close to it

drcrees commented:
Updated warning verbiage, attached is an example of what it will look like.

<details>
<summary>Example Warning Output with New Verbiage</summary>

![warning](https://user-images.githubusercontent.com/11447165/130150181-9943fe51-4610-4c0b-9a81-d01505810109.png)

</details>

> I am hammering on the Keyboard furiously working on it. My target is before the end of the month or very close to it

👍  No rush, so long as these new resources make it in eventually and we can start using an official release.
It's not terribly fun having to cherry pick commits or go through and carefully pull in changes manually.

Happy to contribute!

chb0github commented:
I commited on another PR to release 2.3.x by EOM come hell-or-highwater. I am going to change this pr to merge with 2.3.x and approve it and merge it

drcrees commented:
@chb0github I will resolve conflicts

chb0github commented:
thank you @drcrees - If you have any questions I will be on the gitter channel

drcrees commented:
@chb0github appears you are currently changing how clients are being constructed.  It took me a bit, but I was able to stitch something together locally to get acceptance tests somewhat running again.  They are at least passing for these new resources.

Worth mentioning,

`github.com/jfrog/jfrog-client-go/http/jfroghttpclient`
vs 
`artifactorynew "github.com/jfrog/jfrog-client-go/artifactory"`

These new security resources currently rely on `artifactorynew`.  
I don't know what all you have planned in terms of refactoring `provider.go`, but just thought I would bring it to your attention. since it doesn't look like anything else is currently using `ArtNew`.

chb0github commented:
It's definitely a pivot - thanks for doing the resolution. The
jfrog-client-go actually has some incompatibilities with how terraform
needs to behave and how the current TF provider works. Some places it would
work fine and in some places, the jf-go-client literally makes
contradictory decisions to what the original TF provider did/does. I made a
decision to go with consistency and take more of the decision making into
the side of the TF provider rather than sometimes do it one way, and
sometimes another.

I apologize for the confusion. Thanks again

On Thu, Aug 19, 2021 at 5:42 PM Dylan Crees ***@***.***>
wrote:

> @chb0github <https://github.com/chb0github> appears you are currently
> changing how clients are being constructed. It took me a bit, but I was
> able to stitch something together locally to get acceptance tests somewhat
> running again. They are at least passing for these new resources.
>
> Worth mentioning,
>
> github.com/jfrog/jfrog-client-go/http/jfroghttpclient
> vs
> artifactorynew "github.com/jfrog/jfrog-client-go/artifactory"
>
> These new security resources currently rely on artifactorynew.
> I don't know what all you have planned in terms of refactoring provider.go,
> but just thought I would bring it to your attention. since it doesn't look
> like anything else is currently using ArtNew.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jfrog/terraform-provider-artifactory/pull/95#issuecomment-902346201>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALIRUA6ZZ2MQXPTDCWA4KLT5WQILANCNFSM466IDLRA>
> .
>

chb0github commented:
I changed the base branch to `2.3.security` The 2.0 SDK seems to be ok after some tweaks, even though there are now some deprecation warnings. The bigger problem is that there are misc other issues and the oauth code just hangs - need to further test. 

So, for now, I am going to park your changes here. I don't know why GH still things they conflict - all have been fixed

drcrees commented:
@chb0github conflicts resolved, changes shown are a result of `make fmt`

chb0github commented:
I am closing this PR - your changes have been merged into 2.3 which is about to be released

chb0github commented:
@drcrees - Hey, there has been a request to rename a field:

`no_auto_user_creation = false` to `auto_user_creation = true`

As you created this I felt you should have some input. It's such a new feature, I don't think it's such a big deal. Apparently others have started using it and it's caused some confusion

drcrees commented:
Apologies for the delayed response, I can see how that would cause confusion.  The only reason it is called `no_auto_user_creation` is because that's what the name is in Artifactory's official YAML `samlSettings` configuration.

[Supported Artifactory YAML Configurations](https://www.jfrog.com/confluence/display/JFROG/Artifactory+YAML+Configuration#ArtifactoryYAMLConfiguration-Security(Generalsecurity,PasswordPolicy,LDAP,SAML,OAuth,HTTPSSO,Crowd))

```yaml
security:   #Security configuration (LDAP, SAML, Password Policy, ...)
  samlSettings:   #SAML SSO settings
    enableIntegration: false   #When set, SAML integration is enabled and users may be authenticated via a SAML server
    certificate: certificate   #The X.509 certificate that contains the public key. The certificate must contain the public key to allow Artifactory to verify sign-in requests
    emailAttribute:   #If noAutoUserCreation is set to false or an internal user exists, Artifactory will set the user's email to the value in this attribute that is returned by the SAML login XML response.
    groupAttribute:   #The group attribute in the SAML login XML response
    loginUrl: http://someurl.com/login   #The identity provider login URL (when you try to login, the service provider redirects to this URL)
    logoutUrl: http://someurl.com/logout  #The identity provider logout URL (when you try to logout, the service provider redirects to this URL)
    noAutoUserCreation: true   #When set, for every request from a SAML user, the user is temporarily associated with default groups (if such groups are defined), and the permissions for these groups apply. Without automatic user creation, you must manually create the user inside Artifactory to manage user permissions not attached to their default groups. When not set, authenticated users are automatically created in Artifactory.
    serviceProviderName: serviceProvider   #The Artifactory name in the SAML federation
    allowUserToAccessProfile: false   #Auto created users will have access to their profile page and will be able to perform actions such as generate API key
    autoRedirect: false   #When set, clicking on the login link will direct users to the configured SAML login URL
    syncGroups: false   #When set, in addition to the groups the user is already associated with, he will also be associated with the groups returned in the SAML login response. Note that the user's association with the returned groups is not persistent. It is only valid for the current login session.
    verifyAudienceRestriction: True #When set, an additional verification step will be added opposite the SAML server to validate SAML SSO authentication requests. The verifyAudienceRestriction attribute for SAML SSO is set by default in the JFrog Platform for new Artifactory installations. When upgrading from a previous Artifactory release, this parameter is disabled only if SAML was already configured.
```

alexhung commented:
@drcrees In PR https://github.com/jfrog/terraform-provider-artifactory/pull/232 I've improved the attribute description to match the Artifactory one.

martinm82 commented:
I am wondering why the provider actually shows this warning:

```
Warning: Usage of Undocumented Artifactory API Endpoints
│
│   with artifactory_saml_settings.saml,
│   on main.tf line 1, in resource "artifactory_saml_settings" "saml":
│    1: resource "artifactory_saml_settings" "saml" {
│
│ The artifactory_saml_settings resource uses endpoints that are undocumented and may not work with SaaS environments, or may change without notice.
```



Based on https://www.jfrog.com/confluence/display/JFROG/Artifactory+YAML+Configuration#ArtifactoryYAMLConfiguration-Security(GeneralSecurity,PasswordPolicy,LDAP,SAML,OAuth,HTTPSSO,Crowd) we can configure SAML via REST API via the configuration endpoint.

drcrees commented:
@martinm82,

You are correct, the configuration endpoint is documented and is used for `update` and `delete` operations for these resources.  However, for `read` this endpoint cannot be used since it returns an XML dump for the entire system.  

To get around this I used the following undocumented API endpoints (what the warning is referring to),

`<base-url>/artifactory/api/securityconfig`
`<base-url>/artifactory/api/oauth`
`<base-url>/artifactory/api/saml/config`
